mods/weather/init.lua

```diff
@@ -114,22 +103,93 @@ local function update_clouds()
 			speed = {x = n_speedx * 4, z = n_speedz * 4},
 		})
 		-- now adjust the shadow intensity
-		player:set_lighting({ shadows = { intensity = 0.7 * (1 - density) } })
+		player:set_lighting({
+			shadows = { intensity = 0.7 * (1 - density) }
+		})
 	end
 end
 
+local function purge_effects(player)
+	-- reset potentially touched values to their defaults
+	if randomize_clouds then
+		player:set_clouds({
+			density = 0.4,
+			thickness = 16,
+			speed = { x = 0, z = -2 }
+		})
+	end
+	player:set_lighting({
+		shadows = { intensity = 0 }
+	})
+end
 
-local function cyclic_update()
-	update_clouds()
-	minetest.after(CYCLE, cyclic_update)
+-- Define API hooks
+
+-- override to set state for newly joined players
+weather.enable_on_join = true
+
+-- returns bool for whether weather is active
+-- returns nil if player is offline or weather is disabled
+weather.get_enabled = function(player)
+	return playerlist[player]
 end
 
+-- override weather generation for individual player
+weather.set_enabled = function(player, enable)
+	if enable == playerlist[player] then
+		return
+	end
+	playerlist[player] = enable
+	if enable then
+		if randomize_clouds then
+			update_clouds({player})
+		end
```

Comment:
Seems to be missing the `else` for a default shadow intensity for mgv6 and singlenode? If you were to add the `else`, this would be a code duplication, so I suggest making a new function `add_clouds(player)`.

I also now see the purpose of passing a list of players to `update_clouds`. You could consider passing a set instead (conceptually the more reasonable data structure since players may not appear multiple times). You'd have to change your calls for single players to `update_clouds({[player] = true})`. You could just pass the `playerset` inside the cyclic update.